src/redis/cache_impl.go

```diff
@@ -45,23 +50,33 @@ func unitToDivider(unit pb.RateLimitResponse_RateLimit_Unit) int64 {
 // @param descriptor supplies the descriptor to generate the key for.
 // @param limit supplies the rate limit to generate the key for (may be nil).
 // @param now supplies the current unix time.
-// @return the cache key.
+// @return cacheKey struct.
 func (this *rateLimitCacheImpl) generateCacheKey(
-	domain string, descriptor *pb_struct.RateLimitDescriptor, limit *config.RateLimit, now int64) string {
+	domain string, descriptor *pb_struct.RateLimitDescriptor, limit *config.RateLimit, now int64) cacheKey {
 
 	if limit == nil {
-		return ""
+		return cacheKey{
+			key:       "",
+			perSecond: false,
+		}
 	}
 
-	var cacheKey string = domain + "_"
+	var key = domain + "_"
 	for _, entry := range descriptor.Entries {
-		cacheKey += entry.Key + "_"
-		cacheKey += entry.Value + "_"
+		key += entry.Key + "_"
+		key += entry.Value + "_"
 	}
```

Comment:
How often is this method called? If it's only done initially when the descriptors are loaded, then it's fine, but if it's called in the hot path, this is pretty slow and results in a lot of allocations. Should use a pool of byte.Buffer to build the string if that's the case. The [example](https://golang.org/pkg/sync/#Pool) for sync.Pool is pretty much the same.

Reply:
This is executed in the hot path. I'll fix.